File: 3 Algorithms on Graphs/Week 3 Paths in Graphs 1/PA3/bipartite/bipartite.py
#Uses python3

import sys
import queue

def bipartite(adj):
    #write your code here
    color = [-1] * len(adj)
    color[0] = 1
    
    Q = []
    Q.append(0)
    
    while Q:
        u = Q.pop()
        for v in adj[u]:
            if color[v] == -1:
                Q.append(v)
                color[v] = 1 - color[u]
            elif color[u] == color[v]:
                return 0
        
    return 1

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)
    print(bipartite(adj))

File: 3 Algorithms on Graphs/Week 3 Paths in Graphs 1/PA3/bipartite/bipartite.cpp
#include <iostream>
#include <vector>
#include <queue>

using std::vector;
using std::queue;

int bipartite(vector<vector<int> > &adj) {
  //write your code here
  vector<int> color(adj.size(), -1);
  color[0] = 1;
  
  queue<int> Q;
  Q.push(0);
  
  while (!Q.empty()){
      int u = Q.front();
      Q.pop();
      
      for (int i = 0; i < adj[u].size(); i++){
          int v = adj[u][i];
          if (color[v] == -1){
              Q.push(v);
              color[v] = 1 - color[u];
          }
          else if (color[u] == color[v]) return 0;
      }
      
  }
  return 1;
}

int main() {
  int n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y;
    std::cin >> x >> y;
    adj[x - 1].push_back(y - 1);
    adj[y - 1].push_back(x - 1);
  }
  std::cout << bipartite(adj);
}
